### PR TITLE
Rebased version the PR to add support for ConvTranspose layers

### PR DESCRIPTION
**Pull request by jmitrevs**
# Description

This is a rebased version of #644. Please look there for details, but in summary it adds support for both io_stream and io_parallel compilation of Conv1DTranspose and Conv2DTranspose.

## Type of change

- [x] New feature (non-breaking change which adds functionality)

## Tests

Still a to-do.

## Checklist

- [x] I have read the [guidelines for contributing](https://github.com/fastmachinelearning/hls4ml/blob/main/CONTRIBUTING.md).
- [ ] I have commented my code, particularly in hard-to-understand areas.
- [ ] I have made corresponding changes to the documentation.
- [ ] My changes generate no new warnings.
- [x] I have installed and run `pre-commit` on the files I edited or added.
- [ ] I have added tests that prove my fix is effective or that my feature works.

### CODE REVIEW

**Comment by vloncar** [comment]
Why do we need multidimensional weights here? Passing pointers doesn't work as expected? I like the idea, we'll need it in the future for Vitis, but the `keep_dims` approach feels very clunky.

File: hls4ml/model/layers.py [hunk 1]
```diff
@@ -435,6 +446,55 @@ def initialize(self):
...
+        n_filts, kern_width, n_chan = data.shape
+        new_weights = np.zeros((self.attributes['stride_width'], n_filts, self.attributes['trfilt_width'], n_chan))
+        for i_sw in range(self.attributes['stride_width']):
+            for i_fw in range(self.attributes['trfilt_width']):
+                filt_ind = i_sw + (self.attributes['trfilt_width'] - i_fw - 1) * self.attributes['stride_width']
+                for i_nf in range(n_filts):
+                    for i_nc in range(n_chan):
+                        if filt_ind < kern_width:
+                            new_weights[i_sw][i_nf][i_fw][i_nc] = data[i_nf][filt_ind][i_nc]
+        data = new_weights
```
Comment (jmduarte) [inline hls4ml/model/layers.py:488]: @vloncar I think the "need" for the multidimensional weights comes from this transpose. I think this can still be done even if all the weights are kept single dimensional, but maybe it's less "clean" or "obvious" what's going on?

Comment (jmduarte) [reply hls4ml/model/layers.py:488]: and @Jonathan-Shoemaker's reply: 

https://github.com/fastmachinelearning/hls4ml/pull/644#issuecomment-1664712032

> > What is the meaning of "keep_dims"?
> 
> keep_dims keeps the weight matrix from being entirely flattened, keeping the first "keep_dims" dimensions as is, flattening along all other dimensions.
> 
> The reason for this is that the ConvTranspose is computed as the interleaving of "stride" number of Conv layers. The dimensions kept are for indexing into these different Conv layers. The idea was that the weight matrix of a ConvTranspose layer can be thought of as a disjoint set of weight matrices for Conv layers and treating it as such was easier.

Comment (Jonathan-Shoemaker) [reply hls4ml/model/layers.py:488]: You want to essentially consistently use disjoint submatrices of the weight matrix for sub-computations, and interleave the results from these sub-computations. My understanding is the alternative would be to pass large slices of the matrix. I can't remember if I tested this, but assume it would work, I think I just found it messier.

**Comment by jmitrevs** [comment]
I pushed a first version of a test and tried to make initial fixes, though there are many remaining errors.

**Comment by jmitrevs** [comment]
Looking at the new test, the results see to be:

```
y_keras.shape=(10, 7, 7, 4)
y_hls4ml.shape=(10, 100)
```
and `model_default_t w2[1][1][108];` seem to be the weight dimensions. Will need to investigate further.

**Comment by jmitrevs** [comment]
The output dimensions inconsistency is for padding = "valid". For "same" the output is flattened but there are the correct number of values. However, they still don't match.

**Comment by jmitrevs** [comment]
I am having trouble getting this to work. I wrote a python implementation of conv2d_transpose that works and I think I understand the im2col style io_parallel implementation that's here, but it's not coming together. The keep_dim weights that are written out actually look strangely reordered. I may branch and try to use the code here while dropping the keep_dim concept.

**Comment by Jonathan-Shoemaker** [comment]
> I am having trouble getting this to work. I wrote a python implementation of conv2d_transpose that works and I think I understand the im2col style io_parallel implementation that's here, but it's not coming together. The keep_dim weights that are written out actually look strangely reordered. I may branch and try to use the code here while dropping the keep_dim concept.

One of the purposes of keep_dims is that the weights are purposefully written in a different order than in the original order of the conv_transpose weight matrix. Purpose of this is that the kernel of conv transpose is better computed as separate kernels of smaller conv layers. These kernels are within the transpose kernel but are not continuous in the normal transpose matrix output.

**Comment by jmitrevs** [comment]
With these few changes conv2dtranspose seems to work for io_parallel. @Jonathan-Shoemaker, do you want to take a look at why the other cases in the pytest are failing?

**Comment by jmitrevs** [comment]
Actually, the conv1dtranspose io_parallel test was due to the pytest being misconfigured, but the io_stream tests are still failing.

**Comment by jmitrevs** [comment]
I think the issue is with padding. With padding="same" it seems to work well.

**Comment by jmitrevs** [comment]
Also needed is checking for valid reuse factors for conv transpose.
